composition-js/src/merging/merge.ts

```diff
@@ -3123,27 +3134,31 @@ class Merger {
                   const implementations = intfToImplMap.get(objectName);
                   if (implementations) {
                     for (const impl of implementations) {
-                      const key = `${impl}.${candidate.name}_${accessControlDirective.name}`;
-                      let additionalSources = this.__accessControlAdditionalSources.get(key);
-                      if (!additionalSources) {
-                        additionalSources = [];
+                      const key = `${impl}.${candidate.name}_${name}`;
+                      this.recordAdditionalSource(sourceMap, key, candidate);
+
+                      const otherInterfaces = implToIntfMap.get(impl);
+                      if (otherInterfaces) {
+                        // we now need to propagate field access control upwards from @interfaceObject fields to any
+                        // other interfaces implemented by the given type
+                        for (const intf of otherInterfaces) {
+                          if (objectName == intf) {
+                            // skip current @interfaceObject
+                            continue;
+                          }
+                          const key = `${intf}.${candidate.name}_${name}`;
+                          this.recordAdditionalSource(sourceMap, key, candidate);
+                        }
                       }
```

Comment:
It's possible that some of the object fields implement the same other interface fields, in which case this code will iterate through the same other interfaces more the once.

To fix this, you could:
1. While iterating through the `for` loop of `for (const impl of implementations)`, accumulate other interfaces/`implToIntfMap.get(impl)` into a `Set`.
2. After that iteration, remove the current interface object name/`objectName` from the `Set`.
3. Iterate through the interfaces of the `Set`, creating the key for the interface field and calling `this.recordAdditionalSource()`.